### ISSUE TITLE
Overriding original exceptions in ASP.NET MVC

### ISSUE THREAD

**Issue by tkirill** [bugsnag/bugsnag-dotnet#41]
TL;DR

When [HandleErrorAttribute](https://msdn.microsoft.com/en-us/library/system.web.mvc.handleerrorattribute%28v=vs.118%29.aspx?f=255&MSPPError=-2147217396) isn't used in the project, Bugsnag overrides original exceptions by exception about not found error view.

## Description ##

We found this exception in our logs:

    System.InvalidOperationException: The view 'Error' or its master was not found or no view engine supports the searched locations. The following locations were searched:
    ~/Views/Purchase/Error.aspx
    ~/Views/Purchase/Error.ascx
    ~/Views/Shared/Error.aspx
    ~/Views/Shared/Error.ascx
    ~/Views/Purchase/Error.cshtml
    ~/Views/Purchase/Error.vbhtml
    ~/Views/Shared/Error.cshtml
    ~/Views/Shared/Error.vbhtml
       at System.Web.Mvc.ViewResult.FindView(ControllerContext context)
       at System.Web.Mvc.ViewResultBase.ExecuteResult(ControllerContext context)
       at System.Web.Mvc.Async.AsyncControllerActionInvoker.<>c__DisplayClass21.<BeginInvokeAction>b__1e(IAsyncResult asyncResult)
       at System.Web.Mvc.Controller.<BeginExecuteCore>b__1d(IAsyncResult asyncResult, ExecuteCoreState innerState)
       at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncVoid`1.CallEndDelegate(IAsyncResult asyncResult)
       at System.Web.Mvc.Controller.EndExecuteCore(IAsyncResult asyncResult)
       at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncVoid`1.CallEndDelegate(IAsyncResult asyncResult)
       at System.Web.Mvc.MvcHandler.<BeginProcessRequest>b__5(IAsyncResult asyncResult, ProcessRequestState innerState)
       at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncVoid`1.CallEndDelegate(IAsyncResult asyncResult)
       at System.Web.HttpApplication.CallHandlerExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute()
       at System.Web.HttpApplication.CallHandlerExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute()
       at System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously)

Our custom error pages are static HTML so we don't need `HandleErrorAttribute` and we don't add it to `GlobalFilters.Filters`.  However, rendering error view is what `HandleErrorAttribute` does and it was strange that the errors appeared in our logs.

Surprisingly it was `BugsnagExceptionHandler` who tried to render error view.  After we disabled it the error disappears and we could see original exceptions which cause `BugsnagExceptionHandler` to run.

## Steps to reproduce ##

I created minimal MVC 5 project to demonstrate and reproduce the problem: [tkirill/bugsnag-eat-exception-reproduce](https://github.com/tkirill/bugsnag-eat-exception-reproduce).

Steps to reproduce:

1. Put API key in Web.config.
2. Build project.
3. Navigate to the root page "/".
4. Custom 500 page will be displayed.
5. Comment out Bugsnag action filter in Global.asax.cs.
6. Rebuild project.
7. Navigate to the root page "/".
8. Custom 500 page will be displayed again.
9. Open C:\logs\eat-exceptions.log.

In the log file will be two exceptions:

* One about not found error view. This is the case then Bugsnag action filter activated.
* One about original exception -- InvalidOperationException. This is the case then Bugsnag action filter disabled.

### Additional possible problem ###

Likely the same problem occurs when someone use `HandleErrorAttribute` with custom view name set in the corresponding `.View` property.

## Why ##

I believe the call of [base.OnException](https://github.com/bugsnag/bugsnag-dotnet/blob/dc97a1b2a3a20ba65c8762846c5632da59c3ead2/src/Bugsnag/Clients/WebMVCClient.cs#L120) causes the problem.  Source code of `HandleErrorAttribute` [proves](https://github.com/mono/aspnetwebstack/blob/master/src/System.Web.Mvc/HandleErrorAttribute.cs#L92) this assumption.

## What is expected ##

As for me the expected behavior of `BugsnagExceptionHandler` is only reporting exceptions to Bugsnag.  Trying to display error page isn't its responsibility.

**Comment by CodeHex** [comment]
Yep, great spot! Looks like the `base.OnException()` call should be removed. As you say, the Bugsnag handler should just report the exception rather than try to create and display the Error page to also deal with the exception (which the `base.OnException()` does). Another `HandleErrorAttribute()` should be registered if you want to make use of an Error page. Or in your case only use `<customErrors ... />` in the `Web.config`.

It looks like any exceptions being raised are being sent to Bugsnag correctly, but the exception is being overridden after its been sent (when the Error page is trying to be rendered). So hopeful, the Bugsnag events should be ok, with your local logs having the overridden exception.

Will submit a pull request to remove the call. Thanks for the detail and example, really helped to narrow down the problem :)

**Comment by tkirill** [comment]
@CodeHex I believe #42 fixes the problem but I can't check since the new version wasn't published on NuGet.

However, simply removing `base.OnException` causes another problems.  See #43.
